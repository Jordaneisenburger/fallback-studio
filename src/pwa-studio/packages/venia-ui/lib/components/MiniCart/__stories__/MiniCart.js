import React from 'react';
import { storiesOf } from '@storybook/react';

import '../../../index.css';
import { Adapter } from '@magento/venia-drivers';
import store from '../../../store';

import MiniCart from '../miniCart';
import defaultClasses from '../miniCart.css';

const apiBase = new URL('/graphql', location.origin).toString();
const noop = () => {};

const nonConfigurableItem = {
    __typename: 'SimpleProduct',
    image: {
        __typename: 'MediaGalleryEntry',
        disabled: false,
        file: '/v/a/va11-sg_main.jpg',
        label: 'Main',
        position: 1
    },
    item_id: 1,
    name: 'A Non-Configurable Product',
    options: [],
    price: 49,
    product_type: 'simple',
    qty: 1,
    quote_id: '1234',
    sku: 'SKU'
};
const configurableItem = {
    ...nonConfigurableItem,
    __typename: 'ConfigurableProduct',
    image: {
        ...nonConfigurableItem.image,
        file: '/v/t/vt11-rn_main_1.jpg'
    },
    name: 'A Configurable Product',
    options: [
        { value: 'Rain', label: 'Fashion Color' },
        { value: 'M', label: 'Fashion Size' }
    ],
    product_type: 'configurable'
};

const baseProps = {
    beginEditItem: noop,
    cart: {
        details: {
            currency: {
                quote_currency_code: 'USD'
            },
            items: [],
            items_qty: 0
        },
        editItem: null,
        isEditingItem: false,
        isLoading: false,
        isUpdatingItem: false,
        totals: {
            subtotal: 49
        }
    },
    classes: defaultClasses,
    closeDrawer: noop,
    endEditItem: noop,
    isCartEmpty: true,
    isMiniCartMaskOpen: false,
    isOpen: true,
    removeItemFromCart: noop,
    updateItemInCart: noop
};

const stories = storiesOf('MiniCart', module);

stories.add('Loading', () => {
    const props = {
        ...baseProps,
        cart: {
            ...baseProps.cart,
            isLoading: true
        }
    };

    return (
        <Adapter
            apiBase={apiBase}
            apollo={{ link: Adapter.apolloLink(apiBase) }}
            store={store}
        >
            <MiniCart {...props} />
        </Adapter>
    );
});

stories.add('Empty Cart', () => {
    const props = {
        ...baseProps
    };

    return (
        <Adapter
            apiBase={apiBase}
            apollo={{ link: Adapter.apolloLink(apiBase) }}
            store={store}
        >
            <MiniCart {...props} />
        </Adapter>
    );
});

stories.add('Many Items', () => {
    const props = {
        ...baseProps,
        cart: {
            ...baseProps.cart,
            details: {
                ...baseProps.cart.details,
                items: [
                    configurableItem,
                    // Give each of these a new item_id
                    // to avoid React duplicate key warnings.
                    { ...nonConfigurableItem, item_id: 2 },
                    { ...configurableItem, item_id: 3 },
                    { ...nonConfigurableItem, item_id: 4 },
                    { ...configurableItem, item_id: 5 },
                    { ...nonConfigurableItem, item_id: 6 },
                    { ...configurableItem, item_id: 7 }
                ],
                items_qty: 7
            }
        },
        isCartEmpty: false
    };

    return (
        <Adapter
            apiBase={apiBase}
            apollo={{ link: Adapter.apolloLink(apiBase) }}
            store={store}
        >
            <MiniCart {...props} />
        </Adapter>
    );
});

stories.add('Editing', () => {
    const props = {
        ...baseProps,
        cart: {
            ...baseProps.cart,
            editItem: nonConfigurableItem,
            isEditingItem: true,
            isLoading: false
        },
        isCartEmpty: false
    };

    return (
        <Adapter
            apiBase={apiBase}
            apollo={{ link: Adapter.apolloLink(apiBase) }}
            store={store}
        >
            <MiniCart {...props} />
        </Adapter>
    );
});
