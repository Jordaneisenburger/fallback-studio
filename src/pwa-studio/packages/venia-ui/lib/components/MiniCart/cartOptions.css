.root {
    display: grid;
    grid-template-rows: min-content 1fr;
    height: calc(100% - var(--minicart-header-height));
    left: 0;
    position: absolute;
    top: 3.5rem;
    width: 100%;
}

.focusItem {
    box-shadow: 0 1px rgb(var(--venia-border));
    display: grid;
    gap: 2rem;
    grid-auto-flow: column;
    grid-template-columns: 1fr;
    padding: 1rem;
}

.name {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
}

.form {
    overflow: auto;
}

.modal {
    visibility: hidden;
    height: 100%;
    width: 100%;
    background-color: rgb(var(--venia-grey));
    text-align: center;
    position: absolute;
    bottom: 0;
}

.modal_active {
    visibility: visible;
    composes: modal;
    opacity: 0.9;
}

.options {
    font-weight: bold;
}

.quantity {
    border-color: rgb(var(--venia-border));
    border-style: solid;
    border-width: 0 0 1px;
    margin: 0 1.5rem;
    padding-top: 1rem;
    padding-bottom: 1rem;
}

.quantityTitle {
    composes: title from '../../components/ProductOptions/option.css';
}

.save {
    composes: footer from '../../components/Checkout/flow.css';
}
