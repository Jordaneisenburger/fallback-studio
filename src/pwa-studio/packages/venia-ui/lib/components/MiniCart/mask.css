.root_active {
    composes: root_active from '../Mask/mask.css';
    opacity: 0.25;
}
