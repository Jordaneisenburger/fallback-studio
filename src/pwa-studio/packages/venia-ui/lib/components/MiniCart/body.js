import React from 'react';
import { array, bool, func, object, shape, string } from 'prop-types';

import { mergeClasses } from '../../classify';
import LoadingIndicator from '../LoadingIndicator';

import defaultClasses from './body.css';
import EditItem from './editItem';
import EmptyMiniCartBody from './emptyMiniCartBody';
import ProductList from './productList';
import { useBody } from '@magento/peregrine/lib/talons/MiniCart/useBody';

const loadingIndicator = (
    <LoadingIndicator>{`Fetching Cart...`}</LoadingIndicator>
);

const Body = props => {
    const {
        beginEditItem,
        cartItems,
        closeDrawer,
        currencyCode,
        endEditItem,
        isCartEmpty,
        isEditingItem,
        isLoading,
        isUpdatingItem,
        removeItemFromCart,
        updateItemInCart
    } = props;

    const talonProps = useBody({
        beginEditItem,
        endEditItem
    });

    const { editItem, handleBeginEditItem, handleEndEditItem } = talonProps;

    if (isLoading) {
        return loadingIndicator;
    }

    if (isCartEmpty) {
        return <EmptyMiniCartBody closeDrawer={closeDrawer} />;
    }

    if (isEditingItem) {
        return (
            <EditItem
                currencyCode={currencyCode}
                endEditItem={handleEndEditItem}
                isUpdatingItem={isUpdatingItem}
                item={editItem}
                updateItemInCart={updateItemInCart}
            />
        );
    }

    const classes = mergeClasses(defaultClasses, props.classes);
    return (
        <div className={classes.root}>
            <ProductList
                beginEditItem={handleBeginEditItem}
                cartItems={cartItems}
                currencyCode={currencyCode}
                removeItemFromCart={removeItemFromCart}
            />
        </div>
    );
};

Body.propTypes = {
    beginEditItem: func.isRequired,
    cartItems: array,
    classes: shape({
        root: string
    }),
    closeDrawer: func,
    currencyCode: string,
    editItem: object,
    endEditItem: func,
    isCartEmpty: bool,
    isEditingItem: bool,
    isLoading: bool,
    isUpdatingItem: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default Body;
