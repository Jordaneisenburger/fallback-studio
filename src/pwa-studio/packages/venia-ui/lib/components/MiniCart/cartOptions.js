import React, { Suspense } from 'react';
import { array, bool, func, number, shape, string } from 'prop-types';
import { Form } from 'informed';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../classify';
import LoadingIndicator from '../LoadingIndicator';
import Button from '../Button';
import Quantity from '../ProductQuantity';

import { isProductConfigurable } from '@magento/peregrine/lib/util/isProductConfigurable';

import defaultClasses from './cartOptions.css';
import { useCartOptions } from '@magento/peregrine/lib/talons/MiniCart/useCartOptions';

const Options = React.lazy(() => import('../ProductOptions'));

const loadingIndicator = (
    <LoadingIndicator>
        <span>{'Fetching Options...'}</span>
    </LoadingIndicator>
);

const CartOptions = props => {
    const {
        cartItem,
        configItem,
        currencyCode,
        endEditItem,
        isUpdatingItem,
        updateCart
    } = props;

    const talonProps = useCartOptions({
        cartItem,
        configItem,
        endEditItem,
        updateCart
    });

    const {
        itemName,
        itemPrice,
        itemQuantity,
        handleCancel,
        handleSelectionChange,
        handleUpdate,
        handleValueChange,
        isUpdateDisabled
    } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);
    const modalClass = isUpdatingItem ? classes.modal_active : classes.modal;

    const options = isProductConfigurable(configItem) ? (
        <Suspense fallback={loadingIndicator}>
            <section className={classes.options}>
                <Options
                    onSelectionChange={handleSelectionChange}
                    options={configItem.configurable_options}
                    selectedValues={cartItem.options}
                />
            </section>
        </Suspense>
    ) : null;

    return (
        <Form className={classes.root}>
            <div className={classes.focusItem}>
                <span className={classes.name}>{itemName}</span>
                <span className={classes.price}>
                    <Price currencyCode={currencyCode} value={itemPrice} />
                </span>
            </div>
            <div className={classes.form}>
                {options}
                <section className={classes.quantity}>
                    <h2 className={classes.quantityTitle}>
                        <span>Quantity</span>
                    </h2>
                    <Quantity
                        initialValue={itemQuantity}
                        onValueChange={handleValueChange}
                    />
                </section>
            </div>
            <div className={classes.save}>
                <Button onClick={handleCancel}>
                    <span>Cancel</span>
                </Button>
                <Button
                    priority="high"
                    onClick={handleUpdate}
                    disabled={isUpdateDisabled}
                >
                    <span>Update Cart</span>
                </Button>
            </div>
            <div className={modalClass}>
                <LoadingIndicator>Updating Cart</LoadingIndicator>
            </div>
        </Form>
    );
};

CartOptions.propTypes = {
    cartItem: shape({
        item_id: number.isRequired,
        name: string.isRequired,
        price: number.isRequired,
        qty: number.isRequired
    }),
    classes: shape({
        root: string,
        focusItem: string,
        price: string,
        form: string,
        quantity: string,
        quantityTitle: string,
        save: string,
        modal: string,
        modal_active: string,
        options: string
    }),
    configItem: shape({
        __typename: string,
        configurable_options: array
    }).isRequired,
    currencyCode: string,
    endEditItem: func.isRequired,
    isUpdatingItem: bool,
    updateCart: func.isRequired
};

export default CartOptions;
