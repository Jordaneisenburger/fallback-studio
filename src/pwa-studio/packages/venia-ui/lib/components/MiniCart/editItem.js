import React from 'react';
import { bool, func, object, string } from 'prop-types';

import LoadingIndicator from '../LoadingIndicator';
import PRODUCT_DETAILS from '../../queries/getProductDetailByName.graphql';

import CartOptions from './cartOptions';
import { useEditItem } from '@magento/peregrine/lib/talons/MiniCart/useEditItem';

const loadingIndicator = (
    <LoadingIndicator>{`Fetching Item Options...`}</LoadingIndicator>
);

const EditItem = props => {
    const {
        currencyCode,
        endEditItem,
        isUpdatingItem,
        item,
        updateItemInCart
    } = props;

    const talonProps = useEditItem({
        item,
        query: PRODUCT_DETAILS
    });

    const { configItem, hasError, isLoading, itemHasOptions } = talonProps;

    if (hasError) {
        return <span>Unable to fetch item options.</span>;
    }

    // If we are loading, or if we know we have options but haven't received
    // them from the query, render a loading indicator.
    if (isLoading || (itemHasOptions && !configItem)) {
        return loadingIndicator;
    }

    return (
        <CartOptions
            cartItem={item}
            configItem={configItem || {}}
            currencyCode={currencyCode}
            endEditItem={endEditItem}
            isUpdatingItem={isUpdatingItem}
            updateCart={updateItemInCart}
        />
    );
};

EditItem.propTypes = {
    currencyCode: string,
    endEditItem: func,
    isUpdatingItem: bool,
    item: object.isRequired,
    updateItemInCart: func
};

export default EditItem;
