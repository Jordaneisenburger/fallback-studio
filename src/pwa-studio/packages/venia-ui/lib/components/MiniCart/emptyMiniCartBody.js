import React from 'react';
import { func, string, shape } from 'prop-types';

import { mergeClasses } from '../../classify';
import Trigger from '../Trigger';

import defaultClasses from './emptyMiniCartBody.css';
import { useEmptyMiniCart } from '@magento/peregrine/lib/talons/MiniCart/useEmptyMiniCart';

const EmptyMiniCart = props => {
    const { closeDrawer } = props;

    const talonProps = useEmptyMiniCart({
        closeDrawer
    });

    const { handleClick } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);

    return (
        <div className={classes.root}>
            <h3 className={classes.emptyTitle}>
                There are no items in your shopping cart
            </h3>
            <Trigger action={handleClick}>
                <span className={classes.continue}>Continue Shopping</span>
            </Trigger>
        </div>
    );
};

EmptyMiniCart.propTypes = {
    classes: shape({
        root: string,
        emptyTitle: string,
        continue: string
    }),
    closeDrawer: func
};

export default EmptyMiniCart;
