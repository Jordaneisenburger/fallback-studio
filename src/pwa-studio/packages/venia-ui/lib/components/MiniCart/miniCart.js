import React from 'react';
import { shape, string } from 'prop-types';

import Body from './body';
import Footer from './footer';
import Header from './header';
import Mask from './mask';
import defaultClasses from './miniCart.css';

import { mergeClasses } from '../../classify';
import { useMiniCart } from '@magento/peregrine/lib/talons/MiniCart/useMiniCart';

const MiniCart = props => {
    const {
        cartItems,
        cartState,
        currencyCode,
        handleBeginEditItem,
        handleDismiss,
        handleEndEditItem,
        handleClose,
        handleUpdateItemInCart,
        isEditingItem,
        isLoading,
        isMiniCartMaskOpen,
        isOpen,
        isUpdatingItem,
        numItems,
        removeItemFromCart,
        setStep,
        shouldShowFooter,
        step,
        subtotal
    } = useMiniCart();

    const footer = shouldShowFooter ? (
        <Footer
            currencyCode={currencyCode}
            isMiniCartMaskOpen={isMiniCartMaskOpen}
            numItems={numItems}
            setStep={setStep}
            step={step}
            subtotal={subtotal}
        />
    ) : null;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;

    return (
        <aside className={rootClass}>
            <Header closeDrawer={handleClose} isEditingItem={isEditingItem} />
            <Body
                beginEditItem={handleBeginEditItem}
                cartItems={cartItems}
                closeDrawer={handleClose}
                currencyCode={currencyCode}
                endEditItem={handleEndEditItem}
                isCartEmpty={cartState.isEmpty}
                isEditingItem={isEditingItem}
                isLoading={isLoading}
                isUpdatingItem={isUpdatingItem}
                removeItemFromCart={removeItemFromCart}
                updateItemInCart={handleUpdateItemInCart}
            />
            <Mask isActive={isMiniCartMaskOpen} dismiss={handleDismiss} />
            {footer}
        </aside>
    );
};

MiniCart.propTypes = {
    classes: shape({
        header: string,
        root: string,
        root_open: string,
        title: string
    })
};

export default MiniCart;
