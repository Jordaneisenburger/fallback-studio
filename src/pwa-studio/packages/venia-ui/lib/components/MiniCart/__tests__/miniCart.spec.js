import React from 'react';
import { act } from 'react-test-renderer';
import { createTestInstance } from '@magento/peregrine';

import MiniCart from '../miniCart';
import Body from '../body';
import Footer from '../footer';
import { useCartContext } from '@magento/peregrine/lib/context/cart';

jest.mock('../body', () => 'Body');
jest.mock('../footer', () => 'Footer');
jest.mock('@magento/peregrine/lib/context/app', () => {
    const state = {};
    const api = { closeDrawer: jest.fn() };
    const useAppContext = jest.fn(() => [state, api]);

    return { useAppContext };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = {
        details: {},
        totals: {}
    };
    const api = { updateItemInCart: jest.fn(), removeItemFromCart: jest.fn() };
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('@magento/peregrine/lib/context/checkout', () => {
    const state = {};
    const api = { cancelCheckout: jest.fn() };
    const useCheckoutContext = jest.fn(() => [state, api]);

    return { useCheckoutContext };
});

const baseProps = {
    cancelCheckout: jest.fn(),
    isOpen: true
};

test('renders the correct tree', () => {
    const [cartState, cartApi] = useCartContext();
    useCartContext.mockReturnValueOnce([
        {
            ...cartState,
            details: {
                currency: {
                    quote_currency_code: 'NZD'
                },
                items: [
                    {
                        item_id: 1,
                        name: 'Unit Test Item',
                        price: 99,
                        product_type: 'configurable',
                        qty: 1,
                        quote_id: '1234',
                        sku: 'SKU'
                    }
                ],
                items_qty: 1
            },
            editItem: null,
            isEditingItem: false,
            isLoading: false,
            isUpdatingItem: false,
            totals: {
                subtotal: 99
            }
        },
        cartApi
    ]);

    const instance = createTestInstance(<MiniCart {...baseProps} />);

    expect(instance.toJSON()).toMatchSnapshot();
});

test('doesnt render a footer when cart is empty', () => {
    const [cartState, cartApi] = useCartContext();
    useCartContext.mockReturnValueOnce([
        {
            ...cartState,
            isEmpty: true
        },
        cartApi
    ]);

    const instance = createTestInstance(<MiniCart {...baseProps} />);
    expect(() => {
        instance.root.findByType(Footer);
    }).toThrow();
});

test('doesnt render a footer when cart is editing', () => {
    const instance = createTestInstance(<MiniCart {...baseProps} />);

    act(() => {
        instance.root.findByType(Body).props.beginEditItem();
    });

    expect(() => {
        instance.root.findByType(Footer);
    }).toThrow();
});

test('doesnt render a footer when cart is loading', () => {
    const [cartState, cartApi] = useCartContext();
    useCartContext.mockReturnValueOnce([
        {
            ...cartState,
            isLoading: true
        },
        cartApi
    ]);

    const instance = createTestInstance(<MiniCart {...baseProps} />);
    expect(() => {
        instance.root.findByType(Footer);
    }).toThrow();
});
