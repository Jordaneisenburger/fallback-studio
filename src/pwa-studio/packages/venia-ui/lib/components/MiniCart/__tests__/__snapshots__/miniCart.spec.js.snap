// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<aside>
  <div>
    <h2>
      Shopping Cart
    </h2>
    <button
      onClick={[Function]}
      type="button"
    >
      <span>
        <svg
          fill="none"
          height="24"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="18"
            x2="6"
            y1="6"
            y2="18"
          />
          <line
            x1="6"
            x2="18"
            y1="6"
            y2="18"
          />
        </svg>
      </span>
    </button>
  </div>
  <Body
    beginEditItem={[Function]}
    cartItems={
      Array [
        Object {
          "item_id": 1,
          "name": "Unit Test Item",
          "price": 99,
          "product_type": "configurable",
          "qty": 1,
          "quote_id": "1234",
          "sku": "SKU",
        },
      ]
    }
    closeDrawer={[Function]}
    currencyCode="NZD"
    endEditItem={[Function]}
    isEditingItem={false}
    isLoading={false}
    isUpdatingItem={false}
    removeItemFromCart={[MockFunction]}
    updateItemInCart={[Function]}
  />
  <button
    onClick={[Function]}
  />
  <Footer
    currencyCode="NZD"
    isMiniCartMaskOpen={false}
    numItems={1}
    setStep={[Function]}
    step="cart"
    subtotal={99}
  />
</aside>
`;
